lib/paper_trail.rb

```diff
@@ -103,59 +138,54 @@ def timestamp_field=(_field_name)
       raise(E_TIMESTAMP_FIELD_CONFIG)
     end
 
-    # Sets who is responsible for any changes that occur. You would normally use
-    # this in a migration or on the console, when working with models directly.
-    # In a controller it is set automatically to the `current_user`.
+    # See PaperTrail::Request for information
     # @api public
     def whodunnit=(value)
-      paper_trail_store[:whodunnit] = value
+      ::ActiveSupport::Deprecation.warn(
+        "PaperTrail.whodunnit= is deprecated, use PaperTrail.request.whodunnit="
+      )
+      request.whodunnit = value
     end
 
-    # If nothing passed, returns who is reponsible for any changes that occur.
-    #
-    #   PaperTrail.whodunnit = "someone"
-    #   PaperTrail.whodunnit # => "someone"
-    #
-    # If value and block passed, set this value as whodunnit for the duration of the block
-    #
-    #   PaperTrail.whodunnit("me") do
-    #     puts PaperTrail.whodunnit # => "me"
-    #   end
-    #
+    # See PaperTrail::Request for information
     # @api public
     def whodunnit(value = nil)
-      if value
-        raise ArgumentError, "no block given" unless block_given?
-
-        previous_whodunnit = paper_trail_store[:whodunnit]
-        paper_trail_store[:whodunnit] = value
-
+      if value.nil?
+        ::ActiveSupport::Deprecation.warn(
+          "PaperTrail.whodunnit is deprecated, use PaperTrail.request.whodunnit"
+        )
+        request.whodunnit
+      else
+        ::ActiveSupport::Deprecation.warn(
+          "Passing a block to PaperTrail.whodunnit is deprecated, " \
+          'use PaperTrail.request(whodunnit: "John") do .. end'
+        )
+        before = request.whodunnit
```

Comment:
Why not use method as suggested in the deprecation warning?

e.g.

```ruby
request(whodunnit: value)
  yield
end
```

Reply:
Good suggestion, thanks.